### ISSUE TITLE
Refactor `utils.report` to not restore mutated nodes in `computeEditInfo`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8364]
See: https://github.com/stylelint/stylelint/blob/19ee47d0a8afde625672a1fa6c76ac3d60015bd8/lib/utils/report.mjs#L272-L289

I added `node.replaceWith(fixedNodeClone);` because I thought it would be cleaner if the final AST matched the original.

But this isn't actually needed for any current feature.

I found that this detail makes it slightly less intuitive to use the `FixObject` API.

```ts
type FixObject = {
	apply?: FixCallback;
	node?: PostCSS.Node;
};
```

A removal of a node currently requires passing the parent instead of the node itself.
A replacement of a node also requires passing the parent.
This isn't obvious and we can't warn users when they get it wrong.

Current requirements:

| scenario | mutated node in `FixObject` |
| --- | ---- |
| edit | node itself |
| replace | parent node |
| remove | parent node |
| sort | parent node |
| add before/after | parent node |

The computed edit info will also be correct even when users pass the node itself instead of the parent.
Only restoring the AST will not work and we don't leverage that anyway.

I suggest that we remove that implementation detail entirely.
This will simplify what users should pass as `node` in `FixObject`:

| scenario | mutated node in `FixObject` |
| --- | ---- |
| edit | node itself |
| replace | node itself |
| remove | node itself |
| sort | parent node |
| add before/after | parent node |

A side effect of dropping this unused feature is that we avoid overlap in more cases and can collect more edit info's in a single run. Anytime the parent is used the range will obviously span a lot more surface. Using the node itself in more scenarios seems better. 

-----

If we do ever need a clean AST at the end of collecting warnings we can better clone the original before doing any work.

------

Thoughts?

**Comment by romainmenke** [comment]
> In these cases, will you throw a custom message if the `parent` is absent?

No.

If `FixObject.node` is missing the edit info won't be computed.
It is equivalent to not providing `FixObject`.

It is up to rule authors to have correct implementations and to have sufficient test coverage.

**Comment by Mouvedia** [comment]
> It is up to rule authors to have correct implementations…

Follow-up question then, if `fix.node` is not provided should we use the `node` passed by to `report` by default?
i.e. it may often be the same IIUC

**Comment by romainmenke** [comment]
Definitely not :)
This will only lead to bugs.

Currently this new feature is opt-in and requires rule authors to specifically refactor to make this work. This nudges them to make specific choices and add test coverage.

This would also be against other efforts to improve `report()` where we are now deprecating ambiguous position arguments.

Allowing the node passed to `report` to be used instead we would introduce new subtle and possibly confusing behavior.

**Comment by Mouvedia** [comment]
This change will require to have 2 examples, instead of 1 previously, in the documentation.
e.g. for _edit_ and _add_

---

The rest can be fleshed out in the PR itself.

**Comment by jeddy3** [comment]
All sounds good to me. Thank you for writing up the proposal in clear detail!
